### ISSUE TITLE
Not all tiles extracted with normalization options enabled

### ISSUE THREAD

**Issue by juanenofbit** [jamesdolezal/slideflow#222]
Hello,

When extracting tiles with normalization options, only a subset of them are extracted (around the edges). 
No changes to the rest of the parameters (thresholds and fractions to ensure that all tiles within the ROI are extracted)
Thanks!

`dataset.extract_tiles(
    normalizer='reinhard',
    whitespace_threshold=255,
    whitespace_fraction=1,
    grayspace_threshold=0,
    grayspace_fraction=1,
    skip_extracted=False,
    save_tiles=True,
    roi_method='auto')`

![Image2](https://user-images.githubusercontent.com/86926226/214877292-2fce5239-ae5a-4f4d-a6cc-0682dc6538bf.png)

Without normalization: 
`dataset.extract_tiles(
    normalizer=None,
    whitespace_threshold=255,
    whitespace_fraction=1,
    grayspace_threshold=0,
    grayspace_fraction=1,
    skip_extracted=False,
    save_tiles=True,
    roi_method='auto')`

![Image1](https://user-images.githubusercontent.com/86926226/214877511-3e82e069-47dd-495b-9f3d-d78420bb9b79.png)

**Comment by jamesdolezal** [comment]
That's a weird one! I'm actually not able to reproduce it on my end - ROIs (with holes) still extract the same tiles, with or without normalization.

Could you kindly share the full slide processing pipeline you are using? It looks like you are using some form of QC (Otsu? Blur?). And are you running from a published version or from the master branch?

**Comment by jamesdolezal** [comment]
Actually - I was able to reproduce the issue. It appears to be specific to the Libvips backend and normalization during tile extraction. An error occurs during normalization that causes the tile to be skipped (see: https://github.com/jamesdolezal/slideflow/blob/3f674f5ff2ff9fdf5519620dc33740f1d249d4d5/slideflow/slide/backends/vips.py#L195). Normalization errors can happen occasionally when tiles are primarily one solid color, although clearly something is wrong here. I will investigate the potential source of the normalization error, and add a debug logging step so tiles are not silently skipped in the future in the case of a normalization error.

**Comment by juanenofbit** [comment]
Thanks for the quick support!

It's quite strange because if I explicitly add "qc='None" (although it is the default value), the number of processed tiles increases but about half of them are still missing.

Is it possible to use another backend besides libvips and avoid this problem?

**Comment by jamesdolezal** [comment]
Ok I found and fixed the issue. It is encountered when:
- Slide backend is Libvips
- Deep learning backend is Tensorflow
- "Reinhard" stain normalization is applied during tile extraction

In the libvips backend, multiprocessing during tile extraction uses the 'spawn' method, which can cause GPU memory issues for processes that use Tensorflow algorithms on GPU. The Reinhard stain normalization has a native Tensorflow implementation, so during tile extraction, each multiprocessing worker was each trying to access the GPU simultaneously, causing the error.

This does not happen in the cuCIM backend, which uses the 'fork' multiprocessing method.

The permanent solution in 32fbf96 is to disable native Tensorflow normalization during tile extraction with the libvips backend.

**Workarounds in the meantime could include:**
- Use cuCIM slide reading backend
- Use PyTorch deep learning backend
- Disable GPU access during tile extraction by setting the environmental variable `CUDA_VISIBLE_DEVICES=-1`

We didn't notice this bug earlier for a couple of reasons:
- Most of the time we extract tiles with GPU access disabled
- We typically prefer normalization in real-time rather than during tile extraction
- The Libvips backend was recently switched to the 'spawn' method since it was discovered that Libvips itself is memory-unsafe with forking
